### PR TITLE
Moved api.go from dev branch to pull request branch.

### PR DESCRIPTION
**Pull request by jlm123hi**
Everything here is either google cloud storage code, and not testable, or support for the tests.

---
This change is [Reviewable](https://reviewable.io/reviews/m-lab/downloader/2)

### CODE REVIEW

**Comment by gfr10598** [comment]
Adding Ya, to make sure she also sees the conversation.

**Comment by gfr10598** [comment]
Could you add a trivial .travis.yml, so that travis is happy?

**Comment by gfr10598** [comment]
Review status: 0 of 1 files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---

*[api.go, line 1 at r1](https://reviewable.io:443/reviews/m-lab/downloader/2#-KqO2Xm85pbC_6YtX117:-KqO2Xm85pbC_6YtX118:b-v3ohjz) ([raw file](https://github.com/m-lab/downloader/blob/00b9b40ce37b06fa0033eaf221be2494ab0eecec/api.go#L1)):*
> ```Go
> package main
> ```

Add an issue to introduce package generally, and TODO to change the package name of this particular file.  In general, I'd like you both to use package sparingly, but enough to prompt some thinking about what should be exported and not, and which package each struct should be placed in.

Actually, how long would it take you to adapt the other files if you put this in package api, in its own directory?

---

*[api.go, line 128 at r1](https://reviewable.io:443/reviews/m-lab/downloader/2#-KqTsGuQjEkhEZrCgSno:-KqTsGuQjEkhEZrCgSnp:b72j41p) ([raw file](https://github.com/m-lab/downloader/blob/00b9b40ce37b06fa0033eaf221be2494ab0eecec/api.go#L128)):*
> ```Go
> 
> //// Obj struct implements both the attrs and the object interfaces for testing
> type obj struct {
> ```

This needs a more descriptive name.  Is it used outside this file?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/m-lab/downloader/2)*

**Comment by gfr10598** [comment]
Review status: 0 of 1 files reviewed at latest revision, 7 unresolved discussions, some commit checks failed.

---

*[api.go, line 20 at r1](https://reviewable.io:443/reviews/m-lab/downloader/2#-KqULm5kOwnce7bPWLkO:-KqULm5kOwnce7bPWLkP:bdm112b) ([raw file](https://github.com/m-lab/downloader/blob/00b9b40ce37b06fa0033eaf221be2494ab0eecec/api.go#L20)):*
> ```Go
> 
> type store interface {
> 	getFile(name string) fileObject
> ```

Looks like fileObject is used for writing, and for getting MD5

---

*[api.go, line 21 at r1](https://reviewable.io:443/reviews/m-lab/downloader/2#-KqULX1k96mEwtAH1xvI:-KqULX1k96mEwtAH1xvJ:b-x7x2ik) ([raw file](https://github.com/m-lab/downloader/blob/00b9b40ce37b06fa0033eaf221be2494ab0eecec/api.go#L21)):*
> ```Go
> type store interface {
> 	getFile(name string) fileObject
> 	getFiles(prefix string) []fileAttributes
> ```

change to listFileAttributes(prefix string) ?

---

*[api.go, line 40 at r1](https://reviewable.io:443/reviews/m-lab/downloader/2#-KqUNoX99NHnIz3KEI8O:-KqUNoXA4VtEG_oBZRST:b-jn31i5) ([raw file](https://github.com/m-lab/downloader/blob/00b9b40ce37b06fa0033eaf221be2494ab0eecec/api.go#L40)):*
> ```Go
> type storeGCS struct {
> 	bkt *storage.BucketHandle
> 	ctx context.Context
> ```

Better to have the context passed to the functions that require it, or created from scratch, rather than carrying around a context with a file object.  Exception is if you have created an object that has a limited lifetime, and need to encapsulate that lifetime with the object.

---

*[api.go, line 72 at r1](https://reviewable.io:443/reviews/m-lab/downloader/2#-KqTvg7Oa0uxLLfpQW1f:-KqTvg7Oa0uxLLfpQW1g:b-scs2fh) ([raw file](https://github.com/m-lab/downloader/blob/00b9b40ce37b06fa0033eaf221be2494ab0eecec/api.go#L72)):*
> ```Go
> }
> 
> func (file *fileObjectGCS) getReader() (io.ReadCloser, error) {
> ```

This is unused?  If so delete for now.

---

*[api.go, line 127 at r1](https://reviewable.io:443/reviews/m-lab/downloader/2#-KqUJq-wQpNbJJdEXtwI:-KqUJq-wQpNbJJdEXtwJ:b-5lmyz3) ([raw file](https://github.com/m-lab/downloader/blob/00b9b40ce37b06fa0033eaf221be2494ab0eecec/api.go#L127)):*
> ```Go
> }
> 
> //// Obj struct implements both the attrs and the object interfaces for testing
> ```

Maybe call this testFileObject, and move it to downloader_test.go ?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/m-lab/downloader/2)*

**Comment by jlm123hi** [comment]
Review status: 0 of 3 files reviewed at latest revision, 7 unresolved discussions.

---

*[api.go, line 1 at r1](https://reviewable.io:443/reviews/m-lab/downloader/2#-KqO2Xm85pbC_6YtX117:-KqUyKMVT_nljoGRpSkg:b-yye7qx) ([raw file](https://github.com/m-lab/downloader/blob/00b9b40ce37b06fa0033eaf221be2494ab0eecec/api.go#L1)):*
<details><summary><i>Previously, gfr10598 (Gregory Russell) wrote…</i></summary><blockquote>

Add an issue to introduce package generally, and TODO to change the package name of this particular file.  In general, I'd like you both to use package sparingly, but enough to prompt some thinking about what should be exported and not, and which package each struct should be placed in.

Actually, how long would it take you to adapt the other files if you put this in package api, in its own directory?
</blockquote></details>

It wouldn't be a huge change. Though I still don't see the benefit of packaging api.go in its own package. 

If you want me to do it, I will.

---

*[api.go, line 20 at r1](https://reviewable.io:443/reviews/m-lab/downloader/2#-KqULm5kOwnce7bPWLkO:-KqUy4-lLvaf3ox0Oxpq:b-896fix) ([raw file](https://github.com/m-lab/downloader/blob/00b9b40ce37b06fa0033eaf221be2494ab0eecec/api.go#L20)):*
<details><summary><i>Previously, gfr10598 (Gregory Russell) wrote…</i></summary><blockquote>

Looks like fileObject is used for writing, and for getting MD5
</blockquote></details>

Done.

---

*[api.go, line 21 at r1](https://reviewable.io:443/reviews/m-lab/downloader/2#-KqULX1k96mEwtAH1xvI:-KqUy8oGr7GbXwcYOc80:b-896fix) ([raw file](https://github.com/m-lab/downloader/blob/00b9b40ce37b06fa0033eaf221be2494ab0eecec/api.go#L21)):*
<details><summary><i>Previously, gfr10598 (Gregory Russell) wrote…</i></summary><blockquote>

change to listFileAttributes(prefix string) ?
</blockquote></details>

Done.

---

*[api.go, line 40 at r1](https://reviewable.io:443/reviews/m-lab/downloader/2#-KqUNoX99NHnIz3KEI8O:-KqUyBYqnaz6oPgqQTkl:b-896fix) ([raw file](https://github.com/m-lab/downloader/blob/00b9b40ce37b06fa0033eaf221be2494ab0eecec/api.go#L40)):*
<details><summary><i>Previously, gfr10598 (Gregory Russell) wrote…</i></summary><blockquote>

Better to have the context passed to the functions that require it, or created from scratch, rather than carrying around a context with a file object.  Exception is if you have created an object that has a limited lifetime, and need to encapsulate that lifetime with the object.
</blockquote></details>

Done.

---

*[api.go, line 72 at r1](https://reviewable.io:443/reviews/m-lab/downloader/2#-KqTvg7Oa0uxLLfpQW1f:-KqUyCE5z4x0wp50y9x0:b-896fix) ([raw file](https://github.com/m-lab/downloader/blob/00b9b40ce37b06fa0033eaf221be2494ab0eecec/api.go#L72)):*
<details><summary><i>Previously, gfr10598 (Gregory Russell) wrote…</i></summary><blockquote>

This is unused?  If so delete for now.
</blockquote></details>

Done.

---

*[api.go, line 127 at r1](https://reviewable.io:443/reviews/m-lab/downloader/2#-KqUJq-wQpNbJJdEXtwI:-KqUyEgpPNaryN4yyxqG:bw9qfzm) ([raw file](https://github.com/m-lab/downloader/blob/00b9b40ce37b06fa0033eaf221be2494ab0eecec/api.go#L127)):*
<details><summary><i>Previously, gfr10598 (Gregory Russell) wrote…</i></summary><blockquote>

Maybe call this testFileObject, and move it to downloader_test.go ?
</blockquote></details>

I'm heading out so I'll fix this tomorrow morning.

---

*[api.go, line 128 at r1](https://reviewable.io:443/reviews/m-lab/downloader/2#-KqTsGuQjEkhEZrCgSno:-KqUyIbi-MlrAzkkUsv8:bw9qfzm) ([raw file](https://github.com/m-lab/downloader/blob/00b9b40ce37b06fa0033eaf221be2494ab0eecec/api.go#L128)):*
<details><summary><i>Previously, gfr10598 (Gregory Russell) wrote…</i></summary><blockquote>

This needs a more descriptive name.  Is it used outside this file?
</blockquote></details>

I'm heading out so I'll fix this tomorrow morning.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/m-lab/downloader/2)*

**Comment by jlm123hi** [comment]
Travis should be happy now.

---

Review status: 0 of 3 files reviewed at latest revision, 7 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/m-lab/downloader/2#-:-KqUyfoa6G7WmDubmilj:b9wdafa)*

**Comment by gfr10598** [comment]
Reviewed 2 of 3 files at r2.
Review status: 2 of 3 files reviewed at latest revision, 10 unresolved discussions.

---

*[api.go, line 1 at r1](https://reviewable.io:443/reviews/m-lab/downloader/2#-KqO2Xm85pbC_6YtX117:-KqUzzqsmx5WE6s8uCEd:b-qkk4pu) ([raw file](https://github.com/m-lab/downloader/blob/00b9b40ce37b06fa0033eaf221be2494ab0eecec/api.go#L1)):*
<details><summary><i>Previously, jlm123hi (Joseph Marques) wrote…</i></summary><blockquote>

It wouldn't be a huge change. Though I still don't see the benefit of packaging api.go in its own package. 

If you want me to do it, I will.
</blockquote></details>

Perfectly fine to leave as is.  I'm mostly looking for opportunity for you to introduce packages, so that we have to think about what is exported, and what is internal to each package.  With just one package, everything is effectively global.

Another alternative, that might make more sense, is to put metrics.go into a separate package.  Sometime in the next few days / PRs, would probably be useful.

---

*[api.go, line 21 at r1](https://reviewable.io:443/reviews/m-lab/downloader/2#-KqULX1k96mEwtAH1xvI:-KqUz3wfh7LGAs5MAHpu:b8zh4p8) ([raw file](https://github.com/m-lab/downloader/blob/00b9b40ce37b06fa0033eaf221be2494ab0eecec/api.go#L21)):*
<details><summary><i>Previously, jlm123hi (Joseph Marques) wrote…</i></summary><blockquote>

Done.
</blockquote></details>

Looks like it didn't stick.  I still see getFiles.

---

*[api.go, line 19 at r2](https://reviewable.io:443/reviews/m-lab/downloader/2#-KqUytl-70gAXxuER8F_:-KqUytl-70gAXxuER8Fa:b-6ehm0h) ([raw file](https://github.com/m-lab/downloader/blob/bb86edcb044da4a6a87accaa729d51036fdabc2c/api.go#L19)):*
> ```Go
> const contextTimeout time.Duration = 2 * time.Minute
> 
> type store interface {
> ```

I think it would be better to call the interface fileStore, and the instances store, basically reversing your current use.  Does that make sense?

---

*[api.go, line 61 at r2](https://reviewable.io:443/reviews/m-lab/downloader/2#-KqUzEpMIESv1d3qqKck:-KqUzEpMIESv1d3qqKcl:bsn3m9j) ([raw file](https://github.com/m-lab/downloader/blob/bb86edcb044da4a6a87accaa729d51036fdabc2c/api.go#L61)):*
> ```Go
> 
> func (file *fileObjectGCS) getWriter() io.WriteCloser {
> 	ctx, _ := context.WithTimeout(context.Background(), contextTimeout)
> ```

Consider in the near future adding a WithDeadline() for each context.

---

*[metrics.go, line 1 at r2](https://reviewable.io:443/reviews/m-lab/downloader/2#-KqUzfP4RmX_sEAK2BNa:-KqUzfP4RmX_sEAK2BNb:b-65ukj) ([raw file](https://github.com/m-lab/downloader/blob/bb86edcb044da4a6a87accaa729d51036fdabc2c/metrics.go#L1)):*
> ```Go
> package main
> ```

I thought metrics.go was coming in a separate PR.  Is this intentional?
The file looks fine, so if you intend to include it here, consider it reviewed.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/m-lab/downloader/2)*

**Comment by gfr10598** [comment]
Review status: 2 of 3 files reviewed at latest revision, 8 unresolved discussions.

---

*[api.go, line 21 at r1](https://reviewable.io:443/reviews/m-lab/downloader/2#-KqULX1k96mEwtAH1xvI:-KqV0LXeSGe56isZ29aN:bu836uw) ([raw file](https://github.com/m-lab/downloader/blob/00b9b40ce37b06fa0033eaf221be2494ab0eecec/api.go#L21)):*
<details><summary><i>Previously, gfr10598 (Gregory Russell) wrote…</i></summary><blockquote>

Looks like it didn't stick.  I still see getFiles.
</blockquote></details>

Hmm, somehow got published before I intended.  Now I see that this returns fileObjects, instead of attributes.  

---

*[api.go, line 21 at r2](https://reviewable.io:443/reviews/m-lab/downloader/2#-KqV1F2tKjFPAcIR9mut:-KqV1F2uwylswti3YNSq:bry5y0q) ([raw file](https://github.com/m-lab/downloader/blob/bb86edcb044da4a6a87accaa729d51036fdabc2c/api.go#L21)):*
> ```Go
> type store interface {
> 	getFile(name string) fileObject
> 	getFiles(prefix string) []fileObject
> ```

Hmm.  What would you think about replacing this with namesToMD5(prefix string), which would return a map[string][]byte, containing the map from file name to MD5?  This would eliminate the need for getName() and getMD5() below, IIUC.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/m-lab/downloader/2)*

**Comment by gfr10598** [comment]
:lgtm:
I'm happy with whatever decisions you make about remaining comments.
Please get one additional LGTM before submitting.

Apologies for not being better organized and linear for these reviews.

---

Reviewed 1 of 3 files at r2.
Review status: all files reviewed at latest revision, 8 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/m-lab/downloader/2#-:-KqV2AlI6VSKLirEgFO0:b-19219z)*

**Comment by jlm123hi** [comment]
Review status: 1 of 3 files reviewed at latest revision, 8 unresolved discussions, some commit checks broke.

---

*[api.go, line 1 at r1](https://reviewable.io:443/reviews/m-lab/downloader/2#-KqO2Xm85pbC_6YtX117:-KqY8zGHQDDE-bwtynVx:b-nxzm3g) ([raw file](https://github.com/m-lab/downloader/blob/00b9b40ce37b06fa0033eaf221be2494ab0eecec/api.go#L1)):*
<details><summary><i>Previously, gfr10598 (Gregory Russell) wrote…</i></summary><blockquote>

Perfectly fine to leave as is.  I'm mostly looking for opportunity for you to introduce packages, so that we have to think about what is exported, and what is internal to each package.  With just one package, everything is effectively global.

Another alternative, that might make more sense, is to put metrics.go into a separate package.  Sometime in the next few days / PRs, would probably be useful.
</blockquote></details>

I like that a lot better, because those really are two distinct parts.

Done.

---

*[api.go, line 127 at r1](https://reviewable.io:443/reviews/m-lab/downloader/2#-KqUJq-wQpNbJJdEXtwI:-KqYRIc8X5eh3K1256wA:b-896fix) ([raw file](https://github.com/m-lab/downloader/blob/00b9b40ce37b06fa0033eaf221be2494ab0eecec/api.go#L127)):*
<details><summary><i>Previously, jlm123hi (Joseph Marques) wrote…</i></summary><blockquote>

I'm heading out so I'll fix this tomorrow morning.
</blockquote></details>

Done.

---

*[api.go, line 128 at r1](https://reviewable.io:443/reviews/m-lab/downloader/2#-KqTsGuQjEkhEZrCgSno:-KqYRJH15CaGdQJ0yTNv:b-896fix) ([raw file](https://github.com/m-lab/downloader/blob/00b9b40ce37b06fa0033eaf221be2494ab0eecec/api.go#L128)):*
<details><summary><i>Previously, jlm123hi (Joseph Marques) wrote…</i></summary><blockquote>

I'm heading out so I'll fix this tomorrow morning.
</blockquote></details>

Done.

---

*[api.go, line 19 at r2](https://reviewable.io:443/reviews/m-lab/downloader/2#-KqUytl-70gAXxuER8F_:-KqY8jxL2tHv6yVNRrLn:b-kfityb) ([raw file](https://github.com/m-lab/downloader/blob/bb86edcb044da4a6a87accaa729d51036fdabc2c/api.go#L19)):*
<details><summary><i>Previously, gfr10598 (Gregory Russell) wrote…</i></summary><blockquote>

I think it would be better to call the interface fileStore, and the instances store, basically reversing your current use.  Does that make sense?
</blockquote></details>

That feels cleaner. I'll do it.

Done.

---

*[api.go, line 21 at r2](https://reviewable.io:443/reviews/m-lab/downloader/2#-KqV1F2tKjFPAcIR9mut:-KqY7zF6Hvd9JtOlwrk3:b-mwb1il) ([raw file](https://github.com/m-lab/downloader/blob/bb86edcb044da4a6a87accaa729d51036fdabc2c/api.go#L21)):*
<details><summary><i>Previously, gfr10598 (Gregory Russell) wrote…</i></summary><blockquote>

Hmm.  What would you think about replacing this with namesToMD5(prefix string), which would return a map[string][]byte, containing the map from file name to MD5?  This would eliminate the need for getName() and getMD5() below, IIUC.
</blockquote></details>

Good idea.

Done.

---

*[api.go, line 61 at r2](https://reviewable.io:443/reviews/m-lab/downloader/2#-KqUzEpMIESv1d3qqKck:-KqY6zRwDYM6pPhJrDIy:bc2150o) ([raw file](https://github.com/m-lab/downloader/blob/bb86edcb044da4a6a87accaa729d51036fdabc2c/api.go#L61)):*
<details><summary><i>Previously, gfr10598 (Gregory Russell) wrote…</i></summary><blockquote>

Consider in the near future adding a WithDeadline() for each context.
</blockquote></details>

WithTimeout is a shorthand for WithDeadline. https://golang.org/pkg/context/#WithTimeout

---

*[metrics.go, line 1 at r2](https://reviewable.io:443/reviews/m-lab/downloader/2#-KqUzfP4RmX_sEAK2BNa:-KqY84gkFhk2_7ZZQtCs:bw9v3yj) ([raw file](https://github.com/m-lab/downloader/blob/bb86edcb044da4a6a87accaa729d51036fdabc2c/metrics.go#L1)):*
<details><summary><i>Previously, gfr10598 (Gregory Russell) wrote…</i></summary><blockquote>

I thought metrics.go was coming in a separate PR.  Is this intentional?
The file looks fine, so if you intend to include it here, consider it reviewed.
</blockquote></details>

This is in fact intentional, the api.go code depends on the metrics.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/m-lab/downloader/2)*

**Comment by gfr10598** [comment]
:lgtm:

---

Reviewed 3 of 3 files at r3.
Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/m-lab/downloader/2#-:-KqYXdqq4CYjRlY7Ndbm:bnfp4nl)*

File: metrics/metrics.go [hunk 1]
```diff
@@ -0,0 +1,60 @@
+package metrics
```
Comment (yachang) [inline metrics/metrics.go]: add a overall summary for metrics.go, like:

// The metrics package defines prometheus metric types and provides
// convenience methods to count ....

**Comment by yachang** [comment]
LGTM

**Pull request merged**